File: clients/go/sync_group_test.go
package volant_test

import (
	"errors"
	"net"
	"testing"
	"time"

	volant "github.com/volant-mq/volant/clients/go"
	"github.com/volant-mq/volant/clients/go/codec"
	"github.com/volant-mq/volant/clients/go/frame"
)

type syncGroupServerResult struct {
	groupID         string
	memberID        string
	generation      uint32
	assignmentBytes []byte
	err             error
}

func serveSyncGroup(t *testing.T, errorCode uint16, assignment []codec.Assignment) (addr string, got *syncGroupServerResult, stop func()) {
	t.Helper()
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	res := &syncGroupServerResult{}
	done := make(chan struct{})
	go func() {
		defer close(done)
		conn, err := ln.Accept()
		if err != nil {
			res.err = err
			return
		}
		defer conn.Close()
		_ = conn.SetDeadline(time.Now().Add(5 * time.Second))
		buf := make([]byte, 0, 4096)
		tmp := make([]byte, 4096)
		for {
			f, rest, err := frame.TryDecode(buf)
			if err != nil {
				res.err = err
				return
			}
			if f == nil {
				n, err := conn.Read(tmp)
				if n > 0 {
					buf = append(buf, tmp[:n]...)
				}
				if err != nil {
					res.err = err
					return
				}
				continue
			}
			buf = append([]byte(nil), rest...)
			if f.Opcode != codec.OpSyncGroup {
				res.err = &frame.ProtocolError{Msg: "unexpected opcode"}
				return
			}
			req, err := codec.DecodeSyncGroupRequest(f.Payload)
			if err != nil {
				res.err = err
				return
			}
			res.groupID = req.GroupID
			res.memberID = req.MemberID
			res.generation = req.Generation
			res.assignmentBytes = req.AssignmentBytes
			asgn := assignment
			if errorCode != 0 {
				asgn = nil
			}
			payload, err := codec.EncodeSyncGroupResponse(codec.SyncGroupResponse{
				ErrorCode:  errorCode,
				Assignment: asgn,
			})
			if err != nil {
				res.err = err
				return
			}
			raw, err := frame.Encode(codec.OpSyncGroupResponse, f.CorrelationID, payload)
			if err != nil {
				res.err = err
				return
			}
			if _, err := conn.Write(raw); err != nil {
				res.err = err
				return
			}
		}
	}()
	return ln.Addr().String(), res, func() {
		_ = ln.Close()
		select {
		case <-done:
		case <-time.After(2 * time.Second):
		}
	}
}

func TestSyncGroupCodecRoundtrip(t *testing.T) {
	req := codec.SyncGroupRequest{
		GroupID: "g1", MemberID: "m1", Generation: 3, AssignmentBytes: []byte{},
	}
	raw, err := codec.EncodeSyncGroupRequest(req)
	if err != nil {
		t.Fatal(err)
	}
	decoded, err := codec.DecodeSyncGroupRequest(raw)
	if err != nil {
		t.Fatal(err)
	}
	if decoded.GroupID != "g1" || decoded.MemberID != "m1" || decoded.Generation != 3 || len(decoded.AssignmentBytes) != 0 {
		t.Fatalf("decoded %+v", decoded)
	}
	resp := codec.SyncGroupResponse{
		ErrorCode:  0,
		Assignment: []codec.Assignment{{Topic: "events", Partition: 2}},
	}
	rraw, err := codec.EncodeSyncGroupResponse(resp)
	if err != nil {
		t.Fatal(err)
	}
	got, err := codec.DecodeSyncGroupResponse(rraw)
	if err != nil {
		t.Fatal(err)
	}
	if got.ErrorCode != 0 || len(got.Assignment) != 1 || got.Assignment[0].Topic != "events" || got.Assignment[0].Partition != 2 {
		t.Fatalf("decoded resp %+v", got)
	}
	dispatched, err := codec.DecodeResponse(codec.OpSyncGroupResponse, rraw)
	if err != nil {
		t.Fatal(err)
	}
	if _, ok := dispatched.(codec.SyncGroupResponse); !ok {
		t.Fatalf("dispatch %T", dispatched)
	}
}

func TestSyncGroupSuccessReturnsAssignment(t *testing.T) {
	asgn := []codec.Assignment{{Topic: "events", Partition: 2}}
	addr, got, stop := serveSyncGroup(t, 0, asgn)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	out, err := c.SyncGroup("g1", "m1", 3)
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.groupID != "g1" || got.memberID != "m1" || got.generation != 3 || len(got.assignmentBytes) != 0 {
		t.Fatalf("wire %+v", got)
	}
	if len(out) != 1 || out[0].Topic != "events" || out[0].Partition != 2 {
		t.Fatalf("assignment %+v", out)
	}
}

func TestSyncGroupUnknownMemberIs10(t *testing.T) {
	addr, _, stop := serveSyncGroup(t, 10, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	_, err = c.SyncGroup("g", "ghost", 1)
	var be *codec.BrokerError
	if !errors.As(err, &be) || be.Code != 10 || be.Op != "sync_group" {
		t.Fatalf("err=%v", err)
	}
}

func TestSyncGroupGenerationMismatchIs9(t *testing.T) {
	addr, _, stop := serveSyncGroup(t, 9, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	_, err = c.SyncGroup("g", "m1", 99)
	var be *codec.BrokerError
	if !errors.As(err, &be) || be.Code != 9 || be.Op != "sync_group" {
		t.Fatalf("err=%v", err)
	}
}

File: clients/java/src/test/java/io/volant/SyncGroupTest.java
package io.volant;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertThrows;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.jupiter.api.Test;

/** SyncGroup client tests against a scripted TCP broker (no live server). */
class SyncGroupTest {
    @Test
    void codecRoundtrip() {
        Codec.SyncGroupRequest req = new Codec.SyncGroupRequest("g1", "m1", 3, new byte[0]);
        byte[] raw = Codec.encodeSyncGroupRequest(req);
        Codec.SyncGroupRequest decoded = Codec.decodeSyncGroupRequest(raw);
        assertEquals("g1", decoded.groupId);
        assertEquals("m1", decoded.memberId);
        assertEquals(3L, decoded.generation);
        assertEquals(0, decoded.assignmentBytes.length);
        Codec.SyncGroupResponse resp = new Codec.SyncGroupResponse(
                0, List.of(new Codec.Assignment("events", 2)));
        byte[] rraw = Codec.encodeSyncGroupResponse(resp);
        Codec.SyncGroupResponse got = Codec.decodeSyncGroupResponse(rraw);
        assertEquals(0, got.errorCode);
        assertEquals(1, got.assignment.size());
        assertEquals("events", got.assignment.get(0).topic);
        assertEquals(2, got.assignment.get(0).partition);
        Object dispatched = Codec.decodeResponse(Codec.OP_SYNC_GROUP_RESPONSE, rraw);
        assertInstanceOf(Codec.SyncGroupResponse.class, dispatched);
    }

    @Test
    void successReturnsAssignment() throws Exception {
        try (SyncGroupServer srv = SyncGroupServer.ok(List.of(new Codec.Assignment("events", 2)))) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                List<Codec.Assignment> got = c.syncGroup("g1", "m1", 3);
                assertEquals(1, got.size());
                assertEquals("events", got.get(0).topic);
                assertEquals(2, got.get(0).partition);
            }
            assertEquals("g1", srv.groupId.get());
            assertEquals("m1", srv.memberId.get());
            assertEquals(3, srv.generation.get());
            assertEquals(0, srv.assignmentBytesLen.get());
        }
    }

    @Test
    void unknownMemberIs10() throws Exception {
        try (SyncGroupServer srv = SyncGroupServer.error(10)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                BrokerException ex = assertThrows(
                        BrokerException.class, () -> c.syncGroup("g", "ghost", 1));
                assertEquals(10, ex.code);
                assertEquals("sync_group", ex.op);
            }
        }
    }

    @Test
    void generationMismatchIs9() throws Exception {
        try (SyncGroupServer srv = SyncGroupServer.error(9)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                BrokerException ex = assertThrows(
                        BrokerException.class, () -> c.syncGroup("g", "m1", 99));
                assertEquals(9, ex.code);
                assertEquals("sync_group", ex.op);
            }
        }
    }

    private static final class SyncGroupServer implements AutoCloseable {
        final int port;
        final AtomicReference<String> groupId = new AtomicReference<>();
        final AtomicReference<String> memberId = new AtomicReference<>();
        final AtomicInteger generation = new AtomicInteger();
        final AtomicInteger assignmentBytesLen = new AtomicInteger();
        private final int errorCode;
        private final List<Codec.Assignment> assignment;
        private final ServerSocket listen;
        private final Thread thread;
        private final AtomicReference<Exception> error = new AtomicReference<>();

        static SyncGroupServer ok(List<Codec.Assignment> assignment) throws IOException {
            return new SyncGroupServer(0, assignment);
        }

        static SyncGroupServer error(int code) throws IOException {
            return new SyncGroupServer(code, List.of());
        }

        private SyncGroupServer(int errorCode, List<Codec.Assignment> assignment) throws IOException {
            this.errorCode = errorCode;
            this.assignment = assignment;
            this.listen = new ServerSocket(0, 1, InetAddress.getByName("127.0.0.1"));
            this.port = listen.getLocalPort();
            this.thread = new Thread(this::serve, "sync-group-stub");
            this.thread.setDaemon(true);
            this.thread.start();
        }

        private void serve() {
            try (Socket conn = listen.accept()) {
                conn.setSoTimeout(5_000);
                InputStream in = conn.getInputStream();
                OutputStream out = conn.getOutputStream();
                byte[] buf = new byte[0];
                while (true) {
                    Frame.Decode d = Frame.tryDecode(buf);
                    if (d.frame == null) {
                        byte[] tmp = new byte[4096];
                        int n = in.read(tmp);
                        if (n < 0) {
                            return;
                        }
                        byte[] next = new byte[buf.length + n];
                        System.arraycopy(buf, 0, next, 0, buf.length);
                        System.arraycopy(tmp, 0, next, buf.length, n);
                        buf = next;
                        continue;
                    }
                    buf = d.rest;
                    if (d.frame.opcode != Codec.OP_SYNC_GROUP) {
                        error.set(new ProtocolException("unexpected opcode " + d.frame.opcode));
                        return;
                    }
                    Codec.SyncGroupRequest req = Codec.decodeSyncGroupRequest(d.frame.payload);
                    groupId.set(req.groupId);
                    memberId.set(req.memberId);
                    generation.set((int) req.generation);
                    assignmentBytesLen.set(req.assignmentBytes.length);
                    List<Codec.Assignment> asgn = errorCode != 0 ? List.of() : assignment;
                    byte[] payload = Codec.encodeSyncGroupResponse(
                            new Codec.SyncGroupResponse(errorCode, asgn));
                    out.write(Frame.encode(
                            Codec.OP_SYNC_GROUP_RESPONSE, d.frame.correlationId, payload));
                    out.flush();
                }
            } catch (Exception e) {
                error.set(e);
            }
        }

        @Override
        public void close() {
            try {
                listen.close();
            } catch (IOException ignored) {
            }
            try {
                thread.join(2_000);
            } catch (InterruptedException ignored) {
                Thread.currentThread().interrupt();
            }
            Exception e = error.get();
            if (e != null) {
                throw new RuntimeException(e);
            }
        }
    }
}
